src/charm.py

```diff
@@ -61,7 +61,7 @@ def __init__(self, *args):
         self._lightkube_field_manager = "lightkube"
         self._profiles_service = None
         self._name = self.model.app.name
-        self._service = "npm start"
+        self._service = "/usr/bin/npm start --prefix /app"
```

Comment:
I believe the rock was [already like that](https://github.com/canonical/kubeflow-rocks/blame/main/centraldashboard/rockcraft.yaml#L18) and I did not bother to change it.

From upstream, the command is `npm start` from the `/app` workdir, the exact is achieved by having the `prefix`.

I don't mind change it, looks good to me either way.

Reply:
If we cannot get the rock to have an identical API, afaict we still ensure the charms accept both docker image and rock interchangably (without code changes).  I **think** that's the case with this PR, too, but because both image and service are changed in the same PR we don't confirm that through CI.  If we want to do it this way, let's separate these so into:
1. change the `service` command (which proves the old docker file accepts it and works)
2. change the image to use the rock (which proves the rock also works with this service command)

This way we know we haven't made a change to `service` that supports only the rock

Reply:
Hey @ca-scribner I have added a [change in the rock](https://github.com/canonical/kubeflow-rocks/pull/76) for not needing to modify the command at all. I'll wait for that to be merged before I update this PR again.

Reply:
Hey @ca-scribner I think we'll need to keep the `--prefix /app` in the Pebble layer of this charm. I am hitting this error:

```
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     raise ChangeError(change.err, change)
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed ops.pebble.ChangeError: cannot perform the following tasks:
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed - Start service "serve" (cannot start service: exited quickly with code 254)
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed ----- Logs from task 0 -----
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed 2024-02-15T10:04:10Z INFO Most recent service output:
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR! code ENOENT
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR! syscall open
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR! path //package.json
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR! errno -2
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR! enoent ENOENT: no such file or directory, open '//package.json'
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR! enoent This is related to npm not being able to find a file.
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR! enoent 
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR! A complete log of this run can be found in:
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed     npm ERR!     /root/.npm/_logs/2024-02-15T10_04_10_474Z-debug-0.log
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed 2024-02-15T10:04:10Z ERROR cannot start service: exited quickly with code 254
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed -----
unit-kubeflow-dashboard-0: 2024-02-15 10:04:10 WARNING unit.kubeflow-dashboard/0.kubeflow-profiles-relation-changed 
```

Which essentially is saying that `npm start` cannot find the `package.json` that would be in `/app`. The `--prefix` essentially forces all scripts to run from within the directory we define.

Reply:
I think the issue is [here](https://github.com/canonical/kubeflow-dashboard-operator/pull/178/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cL162).  The service is `replace`d, so while the rock specifies the `working-dir` it then gets overwritten here.  I think any of these would work:
* switch to `merge`, which should use the original `working-dir`
* keep `replace`, but add the `working-dir` like the rock has

I'm not sure why we have the pattern of replacing these entrypoints now that I think about it.  I wonder if that's a relic from back when we weren't so consistent in our rock-making.  Probably this entrypoint could be changed to just:
```python
layer_config = {
    "summary": "kubeflow-dashboard-operator layer",
    "description": "pebble config layer for kubeflow_dashboard_operator",
    "services": {
        self._pebble_service_name: {
            "override": "merge",
            "environment": {
                "NODE_ENV": "production",
                "USERID_HEADER": "kubeflow-userid",
                "USERID_PREFIX": "",
                "PROFILES_KFAM_SERVICE_HOST": f"{self.profiles_service}.{self.model.name}",
                "REGISTRATION_FLOW": self._registration_flow,
                "DASHBOARD_CONFIGMAP": self._configmap_name,
                "LOGOUT_URL": "/authservice/logout",
                "POD_NAMESPACE": self.model.name,  # Added due to https://github.com/canonical/bundle-kubeflow/issues/698   # noqa E501
            },
        }
    },
}

```

Reply:
> I'm not sure why we have the pattern of replacing these entrypoints now that I think about it. I wonder if that's a relic from back when we weren't so consistent in our rock-making. Probably this entrypoint could be changed to just:

So we had this conversation with @orfeas-k as well and decided to bring it up with the team in this week's rocks sync, but then we did not agree on anything ^_^

I agree that I can just merge the pebble layer, though this has presented issues to me in the past (kfp-viz to be specific). I'll update with both of your suggestions and go with the one that works best. For standardisation, we can discuss with the team later on.

Reply:
I've never had trouble with `merge`, but I might have been lucky :P  The `replace` and adding the working-dir works for me till we decide something else! ty